music21/osmd/osmd.py

```diff
@@ -0,0 +1,233 @@
+# -*- coding: utf-8 -*-
+# -------------------------------------------------------------------------------
+# Name:         osmd/osmd.py
+# Purpose:      display Streams in IPython notebooks using Open Sheet Music Display
+#
+# Authors:      Sven Hollowell
+#
+# Copyright:    Copyright © 2018 Michael Scott Cuthbert and the music21 Project
+# License:      LGPL or BSD, see license.txt
+# -------------------------------------------------------------------------------
+'''
+Registers the .show('osmd') converter for use in IPython notebook
+If not in a notebook (e.g. commandline), it will open a browser window with the rendered score
+See OSMD project page here: https://github.com/opensheetmusicdisplay/opensheetmusicdisplay
+'''
+import unittest
+import time
+import random
+import json
+import os
+import importlib
+import tempfile
+import urllib.request
+import webbrowser
+
+from music21.converter.subConverters import SubConverter
+from music21.instrument import Piano
+from music21.common import getSourceFilePath, runningUnderIPython
+from music21 import exceptions21
+from music21.musicxml import m21ToXml
+from music21 import environment
+
+# when using offline=True, sets the disk location where we will cache the OSMD.min.js file:
+environLocal = environment.Environment('osmd.osmd')
+
+class OpenSheetMusicDisplayException(exceptions21.Music21Exception):
+    pass
+
+
+# Determine if we are running inside an IPython notebook
+# if so will render inline - if not we open a new browser window
+try:
+    loader = importlib.util.find_spec('IPython.core.display')
+except ImportError:
+    loader = None
+hasInstalledIPython = loader is not None
+
+
+def getExtendedModules():
+    # pylint: disable=import-outside-toplevel
+    if hasInstalledIPython:
+        from IPython.core.display import display, HTML, Javascript
+    else:
+        def display():
+            raise OpenSheetMusicDisplayException(
+                'OpenSheetMusicDisplay requires IPython to be installed')
```

Comment:
My fears of this being a little too fancy bore out -- doesn't do anything:

My diff to force this to run like IPython on my system and force problems:
```diff
--- a/music21/osmd/osmd.py
+++ b/music21/osmd/osmd.py
@@ -46,9 +46,9 @@ except ImportError:
 hasInstalledIPython = loader is not None
 
 
-def getExtendedModules():
+def getExtendedModules(forceProblem=True):
     # pylint: disable=import-outside-toplevel
-    if hasInstalledIPython:
+    if hasInstalledIPython and not forceProblem:
         from IPython.core.display import display, HTML, Javascript
     else:
         def display():
@@ -108,7 +108,7 @@ class ConverterOpenSheetMusicDisplay(SubConverter):
 
         # generate script to be run on page
         script = self.musicXMLToScript(xml, divId, offline=offline)
-        if in_ipython:
+        if in_ipython or True:
             self.display(self.HTML(f'<div id="{divId}"></div>'))
             self.display(self.Javascript(script))
         else:
```

Results in a lot of `TypeError`s:
```
======================================================================
FAIL: show (__main__.ConverterOpenSheetMusicDisplay)
Doctest: __main__.ConverterOpenSheetMusicDisplay.show
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/doctest.py", line 2204, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for __main__.ConverterOpenSheetMusicDisplay.show
  File "/Users/jwalls/music21/music21/osmd/osmd.py", line 82, in show

----------------------------------------------------------------------
File "/Users/jwalls/music21/music21/osmd/osmd.py", line 88, in __main__.ConverterOpenSheetMusicDisplay.show
Failed example:
    fig_id1 = s.show('osmd')
Exception raised:
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest __main__.ConverterOpenSheetMusicDisplay.show[2]>", line 1, in <module>
        fig_id1 = s.show('osmd')
      File "/Users/jwalls/music21/music21/stream/__init__.py", line 259, in show
        return super().show(fmt=fmt, app=app, **keywords)
      File "/Users/jwalls/music21/music21/base.py", line 2644, in show
        return formatWriter.show(self,
      File "/Users/jwalls/music21/music21/osmd/osmd.py", line 112, in show
        self.display(self.HTML(f'<div id="{divId}"></div>'))
    TypeError: display() takes 0 positional arguments but 1 was given


----------------------------------------------------------------------
```

If the idea is to raise as last as possible, when actually calling show, then the `__init__` could just catch ImportError and set those variables to None, and then in display, check if None and raise then.

Reply:
Ah so you're forcing it to fail? I agree it's well over fanciness-budget, how about just moving it all into the actual call site? Avoids all the `self.display = display` stuff and we can delete the new exception class.

```python 
        # generate script to be run on page
        script = self.musicXMLToScript(xml, divId, offline=offline)
        if in_ipython and hasInstalledIPython:
            # pylint: disable=import-outside-toplevel
            from IPython.core.display import display, HTML, Javascript
            display(HTML(f'<div id="{divId}"></div>'))
            display(Javascript(script))
        else:
```

Reply:
Just trying to exercise the code, not being gleefully evil! Yes, this looks easier to maintain. I don't think we need to be too worried about performance, but if you want to avoid the import-on-every-call overhead, there's a gist here about optimizing lazy imports inside functions using a global variable:

https://wiki.python.org/moin/PythonSpeed/PerformanceTips

Reply:
There's no worries about an import on every call statement on something that is only called on the order of 1-1000 times per user session.  The bottleneck only happens if you're calling the import thousands of times per minute.  

For instance, when music21 was first released, some Python systems did not have weakref module.  For a time, `import weakref` was called within `def activeSite()` and that was a major slowdown because just about everything calls .activeSite.

Reply:
look in the ipython21 directory for various calls to `runningUnderIPython()` -- that's the "one-place" interface to check if IPython is running.